Main.hs

```diff
@@ -0,0 +1,88 @@
+data Lambda
+    = Variable String
+    | Function String Lambda   
+    | Application Lambda Lambda
+
+instance Show Lambda where
+    show (Variable x) = x
+    show (Function p t) = "(λ" ++ p ++ "." ++ show t ++ ")"
+    show (Application t1 t2) = "(" ++ show t1 ++ show t2 ++ ")"
+
+instance Eq Lambda where
+    (Variable x) == (Variable y) = x == y
+    (Function p t) == (Function p' t') = p == p' && t == t'
+    (Application t1 t2) == (Application t1' t2') = t1 == t1' && t2 == t2'
+    _ == _ = False
+
+
+substitution :: Lambda -> Lambda -> Lambda -> Lambda
```

Comment:
It's unnecessary to use `Lambda` when what is expected is a variable. One can call the function like this and get the exception: 

```
ghci> substitution (Function "x" (Variable "x")) (Function "x" (Variable "x")) (Function "x" (Variable "x"))
*** Exception: Main.hs:(19,1)-(24,81): Non-exhaustive patterns in function substitution
```